scripts/autogen-version

```diff
@@ -1,166 +1,147 @@
 # --short a short one. Or alternatively with --header, a version.h file with
 # various constants defined accordingly.
 
-set -e
-
-usage() {
-    echo "usage $(basename "$0") [--short | --header <file> | --cmake | --store <file>] [--git-root <dir>] [<commit>]"
-}
-
-# Make the top-level git repository the current directory.
-cd "$(dirname "$0")"/..
-
-output=long
-
-while [ $# -ge 1 ]; do
-    if [ "$1" = "--short" ]; then
-        output=short
-        shift
-    elif [ "$1" = "--header" ]; then
-        test $# -lt 2 && usage && exit 1
-        output=header
-        dst=$2
-        shift
-        shift
-    elif [ "$1" = "--cmake" ]; then
-        output=cmake
-        shift
-    elif [ "$1" = "--store" ]; then
-        test $# -lt 2 && usage && exit 1
-        output=store
-        dst=$2
-        shift
-        shift
-    elif [ "$1" = "--git-root" ]; then
-        test $# -lt 2 && usage && exit 1
-        cd "$2" || exit
-        shift
-        shift
-    else
-        break
-    fi
-done
-
-if [ $# != 0 ] && [ $# != 1 ]; then
-    usage
-    exit 1
-fi
-
-get_version() {
-    if git rev-parse --git-dir > /dev/null 2>&1; then
-        ref=HEAD
-        describe_arg="--dirty"
-
-        test -n "$1" && test "$1" != "HEAD" && ref="$1" && describe_arg="$1"
-
-        hash=$(git rev-parse --short "${ref}")
-        branch=$(git symbolic-ref --short "${ref}" 2>/dev/null || echo "${hash}")
-        git_version=$(git describe --always --tags --match "v*" "${describe_arg}" | sed 's/^v//' | sed 's/-\([0-9]*\)-g[0-9a-z]*/.\1/g'
-    )
-    elif [ -f ./VERSION ]; then
-        # Read information from VERSION file which contains values for `branch`, `hash`, and `git_version`.
-        . ./VERSION
-    else
-        >&2 echo "Cannot extract version information: neither git repository nor VERSION file present"
-        exit 1
-    fi
-}
+import argparse
+import os
+import re
+import shutil
+import subprocess
+import tempfile
+
 
-get_version "$@"
+def get_version(ref):
+    git_tree = subprocess.run(
+        ["git", "rev-parse", "--git-dir"],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+    )
+    if git_tree.returncode == 0:
+        hash = subprocess.check_output(
+            ["git", "rev-parse", "--short", ref], universal_newlines=True
+        ).strip()
+
+        branch = hash
+        try:
+            branch = subprocess.check_output(
+                ["git", "symbolic-ref", "--short", ref],
+                universal_newlines=True,
+                stderr=subprocess.DEVNULL,
+            ).strip()
+        except:
+            pass
+
+        describe_arg = ref if ref != "HEAD" else "--dirty"
+        raw_git_version = subprocess.check_output(
+            ["git", "describe", "--always", "--tags", "--match", "v*", describe_arg],
+            universal_newlines=True,
+        )
+        git_version = re.sub(
+            r"-([0-9]*)-g[0-9a-z]*", r".\1", re.sub(r"^v", "", raw_git_version)
```

Comment:
This is complicated enough that it needs a comment documenting the intended effect.

That we match trailing `-g...` here seems to break version computation on release tags, e.g.,

```
# In clean checkout of this PR:
$ git tag v9.9.9
$ git checkout v9.9.9
$ ./scripts/autogen-version
Traceback (most recent call last):
  File "/Users/bbannier/src/spicy/build/./scripts/autogen-version", line 88, in <module>
    version, commit, *_ = git_version.split("-")
ValueError: not enough values to unpack (expected at least 2, got 1)
```

Reply:
I'll have to investigate what's going on here, it doesn't seem to break on my end
```
$ git tag v1.33.7
$ git checkout v.1.33.7
$ ./scripts/autogen-version 
1.33.7 (02be372e)
```
Does your git output match mine?
```
$ git describe --always --tags --match "v*"
v1.33.7

without tagging:

$ git describe --always --tags --match "v*"
v1.1.0-dev-61-g02be372e
```
(regex was copied from 3f7dedd41490a36d1a445c7fa5c12bfd9c602e46, `regex.sub()` should behave just like `sed`)

Reply:
I do get similar output.

Are you sure you are invoking this in a clean checkout? The script would e.g., append `--dirty` if there are unstaged changes which where `git describe` would then produce e.g., `v1.33.7-dirty` which can make the splitting below work (around l.88).

Reply:
Whoops, I cloned the repo but issued the command inside the work one.
I guess we can either fix the regex or change line 116 like this, which is not pretty but I guess should work fine:
`version, commit, *_ = git_version.split("-") + [""]`

Reply:
Yes, I think fixing this after the fact is probably okay. You could also go with

```.python
version, commit, *_ = git_version.split('-') if '-' in git_version else (git_version, "")
```

In any case, let's document this edge case in a comment.

Reply:
Went with your version since it makes it easier to understand what's going on